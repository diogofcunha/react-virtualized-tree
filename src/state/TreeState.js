import {getFlattenedTreePaths} from '../selectors/getFlattenedTree';
import {getNodeFromPath} from '../selectors/nodes';

export class State {
  flattenedTree = null;
  tree = null;

  constructor(tree, flattenedTree) {
    this.tree = tree;
    this.flattenedTree = flattenedTree || getFlattenedTreePaths(tree);
  }
}

export const validateState = state => {
  if (!(state instanceof State)) {
    throw new Error(`Expected a State instance but got ${typeof state}`);
  }
};

/**
 * Immutable structure that represents the TreeState.
 */
export default class TreeState {
  /**
   * Given a state, finds a node at a certain row index.
   * @param {State} state - The current state
   * @param {number} index - The visible row index
   * @return {State} An internal state representation
   */
  static getNodeAt = (state, index) => {
    validateState(state);

    const rowPath = state.flattenedTree[index];

    if (!rowPath) {
      throw Error(
        `Tried to get node at row "${index}" but got nothing, the tree are ${state.flattenedTree.length} visible rows`,
      );
    }

    return getNodeFromPath(rowPath, state.tree);
  };

  /**
   * Given a state, finds a node deepness at a certain row index.
   * @param {State} state - The current state
   * @param {number} index - The visible row index
   * @return {number} The node deepness
   */
  static getNodeDeepness = (state, index) => {
    validateState(state);

    const rowPath = state.flattenedTree[index];

    if (!rowPath) {
      throw Error(
        `Tried to get node at row "${index}" but got nothing, the tree are ${state.flattenedTree.length} visible rows`,
      );
    }

    return rowPath.length - 1;
  };

  /**
   * Given a state and an index, finds the number of visible descendants
   * @param {State} state - The current state
   * @param {number} index - The visible row index
   * @return {number} The number of visible descendants
   */
  static getNumberOfVisibleDescendants = (state, index) => {
    const {id} = TreeState.getNodeAt(state, index);

    const {flattenedTree} = state;
    let i;

    for (i = index; i < flattenedTree.length; i++) {
      const path = flattenedTree[i];

      if (!path.some(p => p === id)) {
        break;
      }
    }

    return Math.max(i - 1 - index, 0);
  };

  /**
   * Given a state, gets the tree
   * @param {State} state - The current state
   * @return {Node[]} The tree
   */
  static getTree = state => {
    validateState(state);

    return state.tree;
  };

  /**
   * Creates an instance of state.
   * @param {Node[]} tree - The original tree
   * @return {State} An internal state representation
   */
  static createFromTree = tree => {
    if (!tree) {
      throw Error('A falsy tree was supplied in tree creation');
    }

    if (!Array.isArray(tree)) {
      throw Error('An invalid tree was supplied in creation');
    }

    return new State(tree);
  };
}
