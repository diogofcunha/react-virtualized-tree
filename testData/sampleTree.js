export const COLLAPSED_CHILDREN = {
  id: 3,
  name: 'Leaf 3',
  state: {
    expanded: false,
    favorite: true,
    deletable: true,
  },
  children: [
    {
      id: 'c-3',
      name: 'Leaf 3 Child',
      state: {},
    },
  ],
};

export const EXPANDED_CHILDREN = {
  id: 2,
  name: 'Leaf 2',
  state: {
    expanded: true,
    deletable: true,
  },
  children: [
    COLLAPSED_CHILDREN,
    {
      id: 4,
      name: 'Leaf 4',
    },
  ],
};

export const EXPANDED_NODE_IN_ROOT = {
  id: 0,
  name: 'Leaf 1',
  state: {
    expanded: true,
  },
  children: [
    EXPANDED_CHILDREN,
    {
      id: 5,
      name: 'Leaf 5',
    },
  ],
};

export const COLLAPSED_NODE_IN_ROOT = {
  id: 1,
  name: 'Leaf 6',
  state: {
    expanded: false,
    deletable: true,
  },
  children: [
    {
      id: 6,
      name: 'Leaf 7',
      state: {
        expanded: false,
      },
      children: [
        {
          id: 7,
          name: 'Leaf 8',
        },
        {
          id: 8,
          name: 'Leaf 9',
        },
      ],
    },
    {
      id: 9,
      name: 'Leaf 10',
    },
  ],
};

export const DELETABLE_IN_ROOT = {
  id: 'z',
  name: 'Leaf z',
  state: {
    deletable: true,
    favorite: true,
  },
};

export const DELETABLE_CHILDREN = EXPANDED_CHILDREN;

export const Nodes = [
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
];
