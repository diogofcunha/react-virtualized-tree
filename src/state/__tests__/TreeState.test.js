import TreeState from '../TreeState';
import {Nodes} from '../../../testData/sampleTree';

describe('TreeState', () => {
  describe('createFromTree', () => {
    test('should fail when falsy value is supplied', () => {
      expect(() => TreeState.createFromTree()).toThrowError('A falsy tree was supplied in tree creation');
      expect(() => TreeState.createFromTree('')).toThrowError('A falsy tree was supplied in tree creation');
      expect(() => TreeState.createFromTree(0)).toThrowError('A falsy tree was supplied in tree creation');
      expect(() => TreeState.createFromTree(false)).toThrowError('A falsy tree was supplied in tree creation');
    });

    test('should fail when an invalid value is supplied', () => {
      expect(() => TreeState.createFromTree({})).toThrowError('An invalid tree was supplied in creation');
      expect(() => TreeState.createFromTree('tree')).toThrowError('An invalid tree was supplied in creation');
      expect(() => TreeState.createFromTree(1234)).toThrowError('An invalid tree was supplied in creation');
      expect(() => TreeState.createFromTree(true)).toThrowError('An invalid tree was supplied in creation');
    });

    test('should create state when a valid tree is supplied', () => {
      const {tree, flattenedTree} = TreeState.createFromTree(Nodes);

      expect(tree).toEqual(Nodes);
      expect(flattenedTree).toMatchSnapshot();
    });
  });

  describe('getNodeAt', () => {
    test('should get a for an existing rowId', () => {
      expect(TreeState.getNodeAt(TreeState.createFromTree(Nodes), 0)).toBe(Nodes[0]);
      expect(TreeState.getNodeAt(TreeState.createFromTree(Nodes), 1)).toMatchSnapshot('2nd row');
      expect(TreeState.getNodeAt(TreeState.createFromTree(Nodes), 2)).toMatchSnapshot('3rd row');
      expect(TreeState.getNodeAt(TreeState.createFromTree(Nodes), 6)).toMatchSnapshot('7th row');
    });

    test('should fail with a custom error when supplied rowId does not exist', () => {
      expect(() => TreeState.getNodeAt(TreeState.createFromTree(Nodes), 25)).toThrowErrorMatchingSnapshot();
    });

    test('should fail for when invalid state is supplied', () => {
      expect(() => TreeState.getNodeAt('state', 0)).toThrowError('Expected a State instance but got string');
      expect(() => TreeState.getNodeAt(1225, 0)).toThrowError('Expected a State instance but got number');
      expect(() => TreeState.getNodeAt([], 0)).toThrowError('Expected a State instance but got object');
      expect(() => TreeState.getNodeAt({}, 0)).toThrowError('Expected a State instance but got object');
      expect(() => TreeState.getNodeAt(true, 0)).toThrowError('Expected a State instance but got boolean');
      expect(() => TreeState.getNodeAt(() => {}, 0)).toThrowError('Expected a State instance but got function');
    });
  });

  describe('getTree', () => {
    test('should get a tree', () => {
      expect(TreeState.getTree(TreeState.createFromTree(Nodes))).toEqual(Nodes);
    });

    test('should fail for when invalid state is supplied', () => {
      expect(() => TreeState.getTree('state')).toThrowError('Expected a State instance but got string');
      expect(() => TreeState.getTree(1225)).toThrowError('Expected a State instance but got number');
      expect(() => TreeState.getTree([])).toThrowError('Expected a State instance but got object');
      expect(() => TreeState.getTree({})).toThrowError('Expected a State instance but got object');
      expect(() => TreeState.getTree(true)).toThrowError('Expected a State instance but got boolean');
      expect(() => TreeState.getTree(() => {})).toThrowError('Expected a State instance but got function');
    });
  });

  describe('getNodeDeepness', () => {
    test('should get the correct deepness for existing rowId', () => {
      expect(TreeState.getNodeDeepness(TreeState.createFromTree(Nodes), 0)).toBe(0);
      expect(TreeState.getNodeDeepness(TreeState.createFromTree(Nodes), 1)).toBe(1);
      expect(TreeState.getNodeDeepness(TreeState.createFromTree(Nodes), 2)).toBe(2);
      expect(TreeState.getNodeDeepness(TreeState.createFromTree(Nodes), 3)).toBe(2);
      expect(TreeState.getNodeDeepness(TreeState.createFromTree(Nodes), 6)).toBe(0);
    });

    test('should fail with a custom error when supplied rowId does not exist', () => {
      expect(() => TreeState.getNodeDeepness(TreeState.createFromTree(Nodes), 40)).toThrowErrorMatchingSnapshot();
    });

    test('should fail for when invalid state is supplied', () => {
      expect(() => TreeState.getNodeDeepness('state', 0)).toThrowError('Expected a State instance but got string');
      expect(() => TreeState.getNodeDeepness(1225, 0)).toThrowError('Expected a State instance but got number');
      expect(() => TreeState.getNodeDeepness([], 0)).toThrowError('Expected a State instance but got object');
      expect(() => TreeState.getNodeDeepness({}, 0)).toThrowError('Expected a State instance but got object');
      expect(() => TreeState.getNodeDeepness(true, 0)).toThrowError('Expected a State instance but got boolean');
      expect(() => TreeState.getNodeDeepness(() => {}, 0)).toThrowError('Expected a State instance but got function');
    });
  });

  describe('getNumberOfVisibleDescendants', () => {
    test('should fail for when invalid state is supplied', () => {
      expect(() => TreeState.getNumberOfVisibleDescendants('state', 0)).toThrowError(
        'Expected a State instance but got string',
      );
      expect(() => TreeState.getNumberOfVisibleDescendants(1225, 0)).toThrowError(
        'Expected a State instance but got number',
      );
      expect(() => TreeState.getNumberOfVisibleDescendants([], 0)).toThrowError(
        'Expected a State instance but got object',
      );
      expect(() => TreeState.getNumberOfVisibleDescendants({}, 0)).toThrowError(
        'Expected a State instance but got object',
      );
      expect(() => TreeState.getNumberOfVisibleDescendants(true, 0)).toThrowError(
        'Expected a State instance but got boolean',
      );
      expect(() => TreeState.getNumberOfVisibleDescendants(() => {}, 0)).toThrowError(
        'Expected a State instance but got function',
      );
    });

    test('should get a correct number of descendants for a node with deep descendants', () => {
      expect(TreeState.getNumberOfVisibleDescendants(TreeState.createFromTree(Nodes), 0)).toEqual(4);
    });

    test('should get a correct number of descendants for a node without grand children', () => {
      expect(TreeState.getNumberOfVisibleDescendants(TreeState.createFromTree(Nodes), 1)).toEqual(2);
    });

    test('should get a correct number of descendants for a node with deep descendants', () => {
      expect(TreeState.getNumberOfVisibleDescendants(TreeState.createFromTree(Nodes), 0)).toEqual(4);
    });

    test('should get 0 descendants for a node that does not have any descendants in the root node', () => {
      expect(TreeState.getNumberOfVisibleDescendants(TreeState.createFromTree(Nodes), 6)).toEqual(0);
    });

    test('should get 0 descendants for a node that does not have any descendants', () => {
      expect(TreeState.getNumberOfVisibleDescendants(TreeState.createFromTree(Nodes), 3)).toEqual(0);
    });
  });
});
