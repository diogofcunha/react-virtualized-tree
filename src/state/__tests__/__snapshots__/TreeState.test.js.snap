// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeState createFromTree should create state when a valid tree is supplied 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    0,
    2,
  ],
  Array [
    0,
    2,
    3,
  ],
  Array [
    0,
    2,
    4,
  ],
  Array [
    0,
    5,
  ],
  Array [
    1,
  ],
  Array [
    "z",
  ],
]
`;

exports[`TreeState getNodeAt should fail with a custom error when supplied rowId does not exist 1`] = `"Tried to get node at row \\"25\\" but got nothing, the tree are 7 visible rows"`;

exports[`TreeState getNodeAt should get a for an existing rowId: 2nd row 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "id": "c-3",
          "name": "Leaf 3 Child",
          "state": Object {},
        },
      ],
      "id": 3,
      "name": "Leaf 3",
      "state": Object {
        "deletable": true,
        "expanded": false,
        "favorite": true,
      },
    },
    Object {
      "id": 4,
      "name": "Leaf 4",
    },
  ],
  "id": 2,
  "name": "Leaf 2",
  "state": Object {
    "deletable": true,
    "expanded": true,
  },
}
`;

exports[`TreeState getNodeAt should get a for an existing rowId: 3rd row 1`] = `
Object {
  "children": Array [
    Object {
      "id": "c-3",
      "name": "Leaf 3 Child",
      "state": Object {},
    },
  ],
  "id": 3,
  "name": "Leaf 3",
  "state": Object {
    "deletable": true,
    "expanded": false,
    "favorite": true,
  },
}
`;

exports[`TreeState getNodeAt should get a for an existing rowId: 7th row 1`] = `
Object {
  "id": "z",
  "name": "Leaf z",
  "state": Object {
    "deletable": true,
    "favorite": true,
  },
}
`;

exports[`TreeState getNodeDeepness should fail with a custom error when supplied rowId does not exist 1`] = `"Tried to get node at row \\"40\\" but got nothing, the tree are 7 visible rows"`;
